text/0000-packages-as-optional-namespaces.md

```diff
@@ -0,0 +1,230 @@
+
+A different separator might make more sense. See the [previous section](#slash-as-a-separator) for more on the original proposal of `/` as a separator.
+
+We could continue to use `/` but also use `@`, i.e. have crates named `@foo/bar`. This is roughly what npm does and it seems to work. The `@` would not show up in source code, but would adequately disambiguate crates and features in Cargo.toml and in URLs.
+
+We could perhaps have `foo-*` get autoreserved if you publish `foo`, as outlined in https://internals.rust-lang.org/t/pre-rfc-hyper-minimalist-namespaces-on-crates-io/13041 . I find that this can lead to unfortunate situations where a namespace traditionally used by one project (e.g. `async-*`) is suddenly given over to a different project (the `async` crate). Furthermore, users cannot trust `foo-bar` to be owned by `foo` because the vast number of grandfathered crates we will have.
+
+Triple colons could work. People might find it confusing, but `foo:::bar` evokes Rust paths without being ambiguous.
+
+We could use `~` which enables Rust code to directly name namespaced packages (as `~` is no longer used in any valid Rust syntax). It looks extremely weird, however.
+
+We could use dots (`foo.bar`). This does evoke some similarity with Rust syntax, however there are ambiguities: `foo.bar` in Rust code could either mean "the field `bar` of local/static `foo`" or it may mean "the crate `foo.bar`".
+
+Note that unquoted dots have semantic meaning in TOML, and allowing for unquoted dots would freeze the list of dependency subfields allowed (to `version`, `git`, `branch`, `features`, etc).
+
+
+We could reverse the order and use `@`, i.e. `foo/bar` becomes `bar@foo`. This might be a tad confusing, and it's unclear how best to surface this in the source.
+
+
+## User / org namespaces
+
+Another way to handle namespacing is to rely on usernames and GitHub orgs as namespace roots. This ties `crates.io` strongly to Github -- currently while GitHub is the only login method, there is nothing preventing others from being added.
+
+Furthermore, usernames are not immutable, and that can lead to a whole host of issues.
+
+The primary goal of this RFC is for _project_ ownership, not _org_ ownership, so it doesn't map cleanly anyway.
+
+## Feature Flags
+
+This proposal allows for optional subsystems.  This can be created today with feature flags by adding a dependency as optional and re-exporting it.
+
+Draw backs to feature flags
+- Solutions for documenting feature flags are limited
+- Feature flags can be cumbersome to work with for users
+- A semver breakage in the optional-subsystem crate is a semver breakage in the namespace crate
+- The optional-subsystem crate cannot depend on the namespace crate
+- There is limited tooling for crate authors to test feature combinations especially in workspaces with feature unification and its slow (re-running all tests even if they aren't relevant)
+
+# Prior art
```

Comment:
Additional prior art: Microsoft had to add https://learn.microsoft.com/en-us/nuget/nuget-org/id-prefix-reservation to NuGet, since otherwise a ton of people were publishing things under `Microsoft.` and `System.`.

One interesting thing it allows is delegation of subnamespaces.  If applied here, that could mean that if `serde-` were reserved then they could delegate `serde-json` to a particular team, if they wanted to curate-but-not-own some of the sub-packages.

Reply:
This feels like a bad argument since C# works pretty differently for packages, with the difference between top-level packages and modules not being as clear. Systems which use domain-style namespaces are a different story, since for example it's not clear whether `Category.Thing` is a single dependency by itself, or if `Thing` is part of the `Category` dependency.

In Rust, the top-level name is the dependency, period. So, you can be confident that `serde::json` is a `json` module within the `serde` crate, and `serde_json` is a crate by itself. Reserving the `serde_` prefix for all crates feels like it's solving a problem that the solution created, since the only notion of ownership is for the root `serde` crate itself.

Reply:
the problem is that in rust a crate can't be splitted while still indicating verifiable common ownership of the resulting cluster of crates, and that it will also stay that way.